### PR TITLE
feat(acir)!: Remove `Oracle` opcode

### LINKED ISSUE #298: Remove Oracle Opcode
### Problem

Oracles will now be a concept inside of Brillig, so we can remove it from the top-level ACIR opcodes.

In Brillig, these are renamed to `Foreign_call_results`

### Proposed solution

Remove Oracle Opcode and corresponding code related to it

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #298 

## Summary\*

This PR removes the `Oracle` opcode as it has been superceded by Brillig foreign calls.

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by ludamad — APPROVED** [review]
lgtm so far

File: acvm/src/pwg/mod.rs [hunk 1]
```diff
@@ -376,17 +350,6 @@ mod tests {
         let w_z = Witness(4);
         let w_z_inverse = Witness(5);
         let opcodes = vec![
-            Opcode::Oracle(OracleData {
```
Comment (TomAFrench) [inline acvm/src/pwg/mod.rs:379]: This test is no longer testing what it says it is. We should rework this.

File: acvm/src/pwg/mod.rs [hunk 2]
```diff
@@ -182,9 +157,8 @@ pub fn solve(
             }
         }
         // We have oracles that must be externally resolved
```
Comment (TomAFrench) [inline acvm/src/pwg/mod.rs:159]:
```suggestion
        // We have foreign calls that must be externally resolved
```

File: acvm/src/pwg/mod.rs [hunk 3]
```diff
@@ -35,17 +30,12 @@ pub enum PartialWitnessGeneratorStatus {
...
+    /// to retrieve information from outside of the ACVM.
+    /// The result of the foreign call is inserted into the `Brillig` opcode which made the call using [`UnresolvedBrilligCall::resolve`].
     ///
-    /// Once this is done, the `PartialWitnessGenerator` can be restarted to solve the remaining opcodes.
-    RequiresOracleData {
-        required_oracle_data: Vec<OracleData>,
+    /// Once this is done, the `PartialWitnessGenerator` can be restarted to solve the new set of opcodes.
+    RequiresForeignCall {
         unsolved_opcodes: Vec<Opcode>,
         unresolved_brillig_calls: Vec<UnresolvedBrilligCall>,
```
Comment (TomAFrench) [inline acvm/src/pwg/mod.rs:40]: Can bikeshed this: https://github.com/noir-lang/acvm/pull/363#issuecomment-1587087301

File: acvm/src/pwg/mod.rs [hunk 4]
```diff
@@ -679,7 +626,7 @@ mod tests {
         let solver_status =
             pwg::solve(&backend, &mut witness_assignments, &mut blocks, next_opcodes_for_solving)
                 .expect("should stall on oracle");
-        let PartialWitnessGeneratorStatus::RequiresOracleData { unsolved_opcodes, mut unresolved_brillig_calls, .. } = solver_status else {
+        let PartialWitnessGeneratorStatus::RequiresForeignCall  { unsolved_opcodes, mut unresolved_brillig_calls, .. } = solver_status else {
             panic!("Should require oracle data")
```
Comment (TomAFrench) [inline acvm/src/pwg/mod.rs:630]: remaining references to oracles

**Review by TomAFrench — COMMENTED** [review]
I'm not in a huge rush to merge this but it would be good to clean up this tech debt. Made some extra notes for areas we need to update

File: acvm/src/pwg/mod.rs (hunk 1 above)
Comment (vezenovm) [reply acvm/src/pwg/mod.rs:379]: I think we can delete this test entirely now

File: acvm/src/pwg/mod.rs (hunk 3 above)
Comment (ludamad) [reply acvm/src/pwg/mod.rs:40]: Not sure I prefer the shorter name, what was the issue with the name? I at least see `foreign_call_wait_info` as OK compared to `wait_info`

Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:40]: > what was the issue with the name?

It's a bit redundant as we have a number of instances of `foreign_call.foreign_call_wait_info` which we can shorten to  `foreign_call.wait_info`.

I'm personally not a fan of the name `wait_info` either however as it's a bit nebulous. Why did we not choose `arguments` for this as the name+inputs can just be thought of as the arguments for the foreign call?

**Comment by TomAFrench** [comment]
Marking as ready-for-review as next release will be breaking.

File: acvm/src/pwg/mod.rs (hunk 3 above)
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:41]: Opened an issue for this here: https://github.com/noir-lang/acvm/issues/396

**Pull request merged**